docs(decorators): fix TypeScript inconsistency

emicovi opened the pull request:
## Description
Fixes the TypeScript syntax inconsistency in the Decorators documentation as reported in #6222.

## Changes Made
- **Moved TypeScript-specific content** to `docs/Reference/TypeScript.md` instead of deleting it
- **Updated Decorators.md** to use consistent JavaScript style matching other decorator methods
- **Added comprehensive TypeScript decorators section** with advanced typing examples
- **Preserved all valuable information** while addressing the original inconsistency issue

## Files Changed
- `docs/Reference/Decorators.md`: Updated to JavaScript-focused documentation
- `docs/Reference/TypeScript.md`: Added new "Decorators" section with TypeScript-specific content

## Issue Reference
Fixes #6222

## Addresses Feedback
- Addresses @jean-michelet's concern about deleting valuable content
- Addresses @jsumners's suggestion to move TypeScript info to TypeScript documentation
- Maintains JavaScript consistency as requested in the original issue

## Testing
- [x] Documentation builds without errors
- [x] All code examples are valid JavaScript/TypeScript
- [x] Consistent style with existing decorator methods
- [x] Cross-references work correctly

jean-michelet commented:
Thanks for your PR.

> - Simplified complex TypeScript-specific concepts (module augmentation, OmitThisParameter)
> - Converted all TypeScript code examples to JavaScript for consistency
> - Replaced TypeScript-specific sections with JavaScript-focused alternatives

No, you've just deleted everything, there's an open issue to discusses what to do.
If the author hasn't made a PR request directly, it's probably because he'd like some feedback before doing so.
I suggest to to wait before merging this.

jsumners commented on docs/Reference/Decorators.md:
```diff
@@ -396,124 +396,65 @@ fastify.register(async function (fastify) {
-The `setDecorator<T>` method provides a safe and convenient way to 
-update the value of a `Request` decorator.  
-If the decorator does not exist, a `FST_ERR_DEC_UNDECLARED` error 
+The `setDecorator` method provides a safe and convenient way to
+update the value of a `Request` decorator.
+If the decorator does not exist, a `FST_ERR_DEC_UNDECLARED` error
```
This is surprising. It seems like the method shouldn't ever throw. But that's not related to documentation. A new issue should be opened to investigate why this is the case.

jean-michelet replied:
If the decorator doesn't exist, I think it's best to let the developers know. 

Erros can happen:
- Typos (`req.acount`, `req.accout`)
- The user thinks he has pre-decorated the request with `decorateRequest('user', null)` but the decoration appeared in an unrelated context.

jean-michelet replied:
Declaration and assignment are two different concepts.
In most languages, trying to assign a value to a variable that hasn't been declared results in a compiler or runtime error, I think it is similar to how Fastify "think" decorators.

Decorators should be declared at registration time. For request decorators, they’re bound to the prototype of the `Request` object, and it’s recommended to assign a default value (e.g., `null` for objects) that closely matches its final form.
This is done for performance reasons, as noted in the decorators doc we are editing in this PR.

If you do something like `request.user = { name: 'Jean' }` without having declared the decorator first, it works. But you silently bypass an important Fastify feature and lose the safety and consistency decorators are meant to provide.

Here’s a quick example:

```js
const app = Fastify()

app.addHook('onRequest', async (req) => {
  // This works but the decorator was never declared
  req.user = { name: 'Jean' }

  // If decorators were enforced, this would fail as expected
  // req.setDecorator('user', { name: 'Jean' })
})

app.get('/', async (req) => {
  return req.user
})

const { body } = await app.inject({
  url: '/',
  method: 'GET'
})

console.log(body) // { name: 'Jean' }
```

> “If it doesn't exist, just add it.”

Totally agree - and using `setDecorator` helps surface the issue by throwing when the decorator isn’t declared.

jsumners reviewed (commented):
Looking good to me. But I wonder if there is information being lost that should be present in the actual TypeScript documentation -- https://fastify.dev/docs/latest/Reference/TypeScript/

emicovi commented:
@jean-michelet @jsumners Thank you for the feedback! I've completely revised my approach based on your comments:

**Changes Made:**
- **Moved TypeScript content to proper location**: Added a comprehensive "Decorators" section to `docs/Reference/TypeScript.md` instead of deleting the valuable information
- **Preserved all content**: Nothing was deleted - all TypeScript-specific examples, module augmentation alternatives, and advanced typing features are now properly organized in the TypeScript documentation
- **Maintained JavaScript consistency**: Updated `docs/Reference/Decorators.md` to use consistent JavaScript style matching the rest of the document
- **Added cross-references**: JavaScript docs now point TypeScript users to the appropriate TypeScript documentation section

**Result:**
- JavaScript users get clean, consistent documentation without confusing TypeScript syntax
- TypeScript users get comprehensive advanced examples in the proper location
- No valuable information was lost - it's now better organized

This addresses both the original issue (#6222) and the concerns raised in the review.

jean-michelet commented:
@jsumners

There are several overlapping concepts related to the use of getDecorator. This isn't strictly TypeScript or JavaScript, but rather a mix of both, along with architectural patterns for working with Fastify.

I understand your concern about preserving the documentation reference, and I’m fine with a PR conforming to the goal of Reference. However, I’d really appreciate it if we could find a place to explain what the original documentation was proposing.

Is there a place for a new Guide ?
I think it can be of interest, I had positive reactions on LinkedIn regarding this way of working.

jean-michelet reviewed (commented):
Perhaps the Typescript document should be more like the Reference document, but for types?

The different ways to structure an application with `getDecorator` could be part of a Guide.

These are just suggestions, I am not blocking anything and let the other collaborators decide.

jean-michelet commented on docs/Reference/Decorators.md:
```diff
@@ -366,201 +366,80 @@ Will define the `foo` property on the Fastify instance:
+  registration time
+- **Runtime safety**: Ensure decorators exist before using them, avoiding
+  undefined access
+- **Plugin encapsulation**: Access decorators within specific plugin contexts
```
It's true that there's an advantage in that regard when using `getDecorator` with generic typing to avoid module augmentation, which pollutes `FastifyInstance` (and request, reply) globally.
But encapsulation or not, `getDecorator` will only fail if the decorator is not available in the current scope and gives no information about encapsulation. 

So this line can be deleted.

jean-michelet commented on docs/Reference/Decorators.md:
```diff
@@ -366,201 +366,80 @@ Will define the `foo` property on the Fastify instance:
-For example:
+The `getDecorator` method is useful for:
+
+- **Plugin dependency validation**: Verify that required decorators exist at
+  registration time
+- **Runtime safety**: Ensure decorators exist before using them, avoiding
+  undefined access
```
These two overlaps.

I would go for something like:
```suggestion
* **Dependency validation**: `getDecorator` can be used to check for required decorators at registration time. If any are missing, it fails at boot, ensuring dependencies are available during the request lifecycle.

jean-michelet commented on docs/Reference/TypeScript.md:
````diff
@@ -687,6 +687,144 @@ Or even explicit config on tsconfig
 }
 ```
 
+### Decorators
````
I think it should be part of `### Plugins`, but it need to be rework.

jean-michelet reviewed (commented):
I think I should push a commit to make the changes by myself and gather your feedback.

Fdawgs commented on docs/Reference/TypeScript.md:
```diff
@@ -687,6 +687,144 @@ Or even explicit config on tsconfig
 
+#### Decorators
+
+Fastify's decorator system allows you to extend the Fastify instance, request, and
```
I know the original docs had it, but could you please remove the [second-person "you" usage](https://fastify.dev/docs/latest/Guides/Style-Guide/#when-to-avoid-the-second-person-you-as-the-pronoun) throughout this?

Fdawgs commented on docs/Reference/Decorators.md:
```diff
@@ -366,201 +366,78 @@ Will define the `foo` property on the Fastify instance:
-This approach modifies the Fastify instance globally, which may lead to
-conflicts and inconsistent behavior in multi-server setups or with plugin
-encapsulation.
+> **Note**: For TypeScript users, `getDecorator` supports generic type parameters.
```
```suggestion
> ℹ️ Note: For TypeScript users, `getDecorator` supports generic type parameters.
```

Fdawgs commented on docs/Reference/Decorators.md:
```diff
@@ -366,201 +366,78 @@ Will define the `foo` property on the Fastify instance:
 
-`getDecorator<T>` on Fastify instance verifies that required decorators are
-available at registration time. 
+**Use Cases**
```
```suggestion

```

Fdawgs commented on docs/Reference/Decorators.md:
````diff
@@ -366,201 +366,78 @@ Will define the `foo` property on the Fastify instance:
 ```
 
----
+**Use Cases**
````
```suggestion

```

Fdawgs commented on docs/Reference/TypeScript.md:
```diff
@@ -687,6 +687,144 @@ Or even explicit config on tsconfig
+
+**Bound Functions Inference**
+
+To save time, it's common to infer function types instead of writing them manually:
```
```suggestion
To save time, it is common to infer function types instead of writing them manually:
```

gurgunday reviewed (approved):
lgtm except @jean-michelet's comment

emicovi commented:
everything should be ok now!

emicovi commented:
Thanks for the feedbacks. I think it's ready to go

Fdawgs commented:
Thanks @emicovi, could you sort the linting out please and then this is good to merge. :)

Fdawgs commented:
@emicovi Linting still needs sorting as I said a month ago.

Uzlopak commented:
@Fdawgs 

I fixed linting...
